File: apps/athenaeum/native/AthenaeumDomain/Sources/AthenaeumDomain/Bookmark.swift
import Foundation

// Phase 5 native stage — mirrors `packages/domain/src/bookmark.ts`. See that file's own header
// comment for why `Bookmark` is deliberately the plan's low-complexity Phase 5 companion to
// Calendar: a capture-inbox record with no OAuth/provider dependency at all.

/// Mirrors `bookmark.ts`'s `BookmarkUrl` brand: "an absolute http(s) URL — validated, not
/// normalized." Same validating-wrapper convention as `ShareKeyHash`/`EntityId`/`Email`.
public struct BookmarkUrl: Hashable, Sendable, CustomStringConvertible, ExpressibleByStringLiteral {
    public let rawValue: String

    /// Matches `bookmark.ts`'s `httpUrlPattern = /^https?:\/\/\S+$/i` exactly.
    private static let pattern = #"^https?://\S+$"#

    public static func isValid(_ value: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return false
        }
        let range = NSRange(value.startIndex..<value.endIndex, in: value)
        return regex.firstMatch(in: value, range: range) != nil
    }

    /// See `EntityId.init(validating:)`'s doc comment for why this is labeled, not `init(_:)`.
    public init(validating rawValue: String) throws {
        guard BookmarkUrl.isValid(rawValue) else {
            throw AthenaeumDomainDecodingError.invalidBookmarkUrl(rawValue)
        }
        self.rawValue = rawValue
    }

    public init(stringLiteral value: String) {
        precondition(BookmarkUrl.isValid(value), "BookmarkUrl literal is not an absolute http(s) URL: \(value)")
        self.rawValue = value
    }

    public var description: String { rawValue }
}

extension BookmarkUrl: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let value = try container.decode(String.self)
        guard BookmarkUrl.isValid(value) else {
            throw DecodingError.dataCorruptedError(
                in: container, debugDescription: "BookmarkUrl must be an absolute http(s) URL, got: \(value)"
            )
        }
        self.rawValue = value
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

/// Mirrors `bookmark.ts`'s `Bookmark` — a captured bookmark, `{id, workspaceId, url, title?,
/// capturedAt, linkedNodeId?}`. `Schema.Class`, diffed by `schema-diff.ts`.
public struct Bookmark: Codable, Hashable, Sendable {
    public let id: EntityId
    public let workspaceId: EntityId
    public let url: BookmarkUrl
    public let title: String?
    public let capturedAt: IsoDateTimeString
    public let linkedNodeId: EntityId?

    public init(
        id: EntityId,
        workspaceId: EntityId,
        url: BookmarkUrl,
        title: String? = nil,
        capturedAt: IsoDateTimeString,
        linkedNodeId: EntityId? = nil
    ) {
        self.id = id
        self.workspaceId = workspaceId
        self.url = url
        self.title = title
        self.capturedAt = capturedAt
        self.linkedNodeId = linkedNodeId
    }
}

File: apps/athenaeum/packages/domain/src/bookmark.ts
import * as Schema from "effect/Schema"
import { EntityId, IsoDateTimeString } from "./node.js"

// Phase 5 domain-extension task, item 2: `Bookmark` — per the plan's "Phase 5 — First gatekeeper:
// Google Calendar + Bookmarks... Bookmarks ship alongside (low complexity, validates 'external
// capture → node' before Phase 6's higher-stakes capture flows)." Deliberately the simplest new
// entity this stage adds: a capture-inbox record, not a gatekeeper-backed sync target — nothing
// about `Bookmark` requires OAuth, a provider API, or observer verification, which is exactly why
// the plan schedules it alongside Calendar as a low-complexity companion.

/** An absolute `http`/`https` URL — validated, not normalized (same "validate, don't normalize"
 *  discipline as `EntityId`/`Email`: a caller-side capture flow, e.g. a browser extension or a
 *  share-sheet target, is expected to hand this schema an already-well-formed absolute URL; this
 *  schema's job is only to reject malformed input before it reaches storage, not to canonicalize
 *  it (strip tracking params, resolve redirects, etc.) — that belongs to whatever future capture
 *  service builds on this schema, not the wire type itself). Branded (rather than a bare
 *  `Schema.String`) for the same reason `ShareKeyHash`/`EntityId`/`Email` all are: a bookmark URL
 *  should never be constructible from, or mistaken for, an arbitrary unvalidated string.
 *
 *  A plain regex, matching `EntityId`'s ULID/UUID pattern and `Email`'s address pattern, rather
 *  than the runtime `URL` constructor: this package has zero Cloudflare/Node.js/DOM-specific
 *  dependencies (`domain`'s own `tsconfig.json` deliberately sets `lib: ["ES2022"]`, no `dom`/
 *  `webworker` — `URL` is not in that type universe), the same reason `dev-auth.ts`'s
 *  `crypto.subtle`-based logic lives in `backend`, not here. */
const httpUrlPattern = /^https?:\/\/\S+$/i

export const BookmarkUrl = Schema.String.pipe(
  Schema.filter((value) => httpUrlPattern.test(value), {
    message: () => "BookmarkUrl must be an absolute http(s) URL"
  }),
  Schema.brand("BookmarkUrl")
)
export type BookmarkUrl = typeof BookmarkUrl.Type

/**
 * A captured bookmark, per the task's own `{id, workspaceId, url, title?, capturedAt, linkedNodeId?}`
 * shape. `title` is optional (a capture flow may only have a raw URL at capture time, e.g. a
 * share-sheet target with no page-title metadata available yet — unlike `CalendarEvent.title`,
 * which is always populated because it is copied straight from a provider's own event resource).
 * `linkedNodeId` mirrors `CalendarEvent.linkedNodeId`'s same "external capture → optional
 * companion node the user can annotate" shape (this file's header comment) — kept optional for
 * the identical reason: this schema takes no position on whether a companion node is created
 * eagerly or lazily, only on how one is referenced once it exists.
 */
export class Bookmark extends Schema.Class<Bookmark>("Bookmark")({
  id: EntityId,
  workspaceId: EntityId,
  url: BookmarkUrl,
  title: Schema.optional(Schema.String),
  capturedAt: IsoDateTimeString,
  linkedNodeId: Schema.optional(EntityId)
}) {}
